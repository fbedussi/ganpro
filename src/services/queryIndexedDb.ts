const dbName = 'ganpro'

let db: IDBDatabase | undefined = undefined

export const _openDb = async (dbName: string): Promise<IDBDatabase> => {
  return db
    ? Promise.resolve(db)
    : new Promise((res, rej) => {
        const request = indexedDB.open(dbName)
        request.onerror = () => {
          rej(request.error)
        }
        request.onsuccess = () => {
          db = request.result
          res(request.result)
        }
        request.onupgradeneeded = () => {
          const db = request.result

          db.createObjectStore('projects', { keyPath: 'id', autoIncrement: true })
          const tasksStore = db.createObjectStore('tasks', { keyPath: 'id', autoIncrement: true })
          tasksStore.createIndex('projId', 'projId', { unique: false })
        }
      })
}

export const queryIndexedDb = (entityName: string) => {
  const db = _openDb(dbName)

  type FetchBaseQueryResult<T> = Promise<
    | {
        data: T
        error?: undefined
      }
    | {
        error: DOMException
        data?: undefined
      }
  >

  function queryFn({
    query,
    operation = 'read',
  }: {
    query: any
    operation?: 'create' | 'read' | 'readAll' | 'update' | 'delete'
  }): FetchBaseQueryResult<any> {
    return db
      .then(
        db =>
          new Promise(res => {
            const transaction = db.transaction([entityName], 'readwrite')
            const objectStore = transaction.objectStore(entityName)

            switch (operation) {
              case 'create': {
                const request = objectStore.add(query)
                request.onsuccess = () => {
                  res(request.result)
                }
                break
              }

              case 'read': {
                const request = objectStore.get(query)
                request.onsuccess = () => {
                  res(request.result)
                }
                break
              }

              case 'readAll': {
                let request
                if (query) {
                  const [indexName, indexValue] = Object.entries(query)[0]
                  const index = objectStore.index(indexName)
                  request = index.getAll(indexValue as IDBValidKey)
                } else {
                  request = objectStore.getAll(query)
                }
                request.onsuccess = () => {
                  res(request.result)
                }
                break
              }

              case 'update': {
                const request = objectStore.put(query)
                request.onsuccess = () => {
                  res(query)
                }
                break
              }

              case 'delete': {
                const request = objectStore.delete(query)
                request.onsuccess = () => {
                  res(true)
                }
                break
              }
            }
          }),
      )
      .then(data => {
        return { data }
      })
      .catch(error => {
        return { error }
      })
  }

  return queryFn
}
