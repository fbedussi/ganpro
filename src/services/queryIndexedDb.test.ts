import { _openDb, queryIndexedDb } from './queryIndexedDb'
import { IDBFactory } from 'fake-indexeddb'

beforeAll(() => {
  // eslint-disable-next-line no-global-assign
  indexedDB = new IDBFactory()
})

const DB_NAME = 'dbname'
const ENTITY_NAME = 'projects'

describe('openDb', () => {
  it('returns a promise that resolves to a db instance', async () => {
    expect(await _openDb(DB_NAME, ENTITY_NAME)).toBeInstanceOf(IDBDatabase)
  })

  it('returns the same instance if called multiple times', async () => {
    const db1 = await _openDb(DB_NAME, ENTITY_NAME)
    const db2 = await _openDb(DB_NAME, ENTITY_NAME)
    expect(db1).toBe(db2)
  })

  // TODO: test the error scenario?
})

describe('queryIndexedDb', () => {
  it('takes an entity name and returns a query function', () => {
    expect(queryIndexedDb('projects')).toEqual(expect.any(Function))
  })

  it('returns a query function that takes a query and returns a promise', () => {
    const queryFn = queryIndexedDb('projects')
    expect(queryFn({ query: 1 })).toEqual(expect.any(Promise))
  })

  describe('CRUD', () => {
    let id: IDBValidKey | undefined
    const fakeProject = { name: 'foo' }

    describe('create an entity', () => {
      it('returns the new entity id', async () => {
        const queryFn = queryIndexedDb('projects')
        const result = await queryFn({ query: fakeProject, operation: 'create' })
        id = result.data
        expect(id).toEqual(expect.any(Number))
      })
    })

    describe('read an entity', () => {
      it('returns the entity', async () => {
        const queryFn = queryIndexedDb('projects')
        const { data } = await queryFn({ query: id! })
        expect(data).toEqual({ ...fakeProject, id })
      })
    })

    describe('read all the entities', () => {
      it('returns all the entities', async () => {
        const queryFn = queryIndexedDb('projects')
        const { data } = await queryFn({ query: undefined, operation: 'readAll' })
        expect(data).toEqual([{ ...fakeProject, id }])
      })
    })

    describe('update an entity', () => {
      it('returns the updated entity', async () => {
        const queryFn = queryIndexedDb('projects')
        const { data } = await queryFn({ query: { id: id!, name: 'baz' }, operation: 'update' })
        expect(data).toEqual({ name: 'baz', id })
      })
    })

    describe('delete an entity', () => {
      it('returns true', async () => {
        const queryFn = queryIndexedDb('projects')
        const { data } = await queryFn({ query: id!, operation: 'delete' })
        expect(data).toEqual(true)
      })

      it('deletes the entity', async () => {
        const queryFn = queryIndexedDb('projects')
        const { data } = await queryFn({ query: id! })
        expect(data).toEqual(undefined)
      })
    })
  })

  describe('Error handling', () => {
    it('querying a non existing store', async () => {
      const queryFn = queryIndexedDb('foo')
      const { error } = await queryFn({ query: 1 })
      expect(error).toMatchSnapshot()
    })

    it('Query a bad id', async () => {
      const queryFn = queryIndexedDb('projects')
      const { error } = await queryFn({ query: undefined as any })
      expect(error).toMatchSnapshot()
    })
  })
})
