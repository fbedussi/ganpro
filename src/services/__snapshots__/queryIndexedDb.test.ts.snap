// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`queryIndexedDb Error handling Query a bad id 1`] = `[DataError: Data provided to an operation does not meet requirements.]`;

exports[`queryIndexedDb Error handling querying a non existing store 1`] = `[NotFoundError: No objectStore named foo in this database]`;
